### PR TITLE
Vault gen password

### LINKED ISSUE #16: UserStory[Security]: Generate password for minio from Vault
### Description
As of today we use user provided screts to access minio, however, going forward we would like to generate these secrets from Vault. The [KV Secrets Engine](https://www.vaultproject.io/docs/secrets/kv) feature could be made use of to achieve this.

**Acceptance Criteria**

1. Remove the hardcoded **minio_access_key**, **minio_secret_key** . 
2. Generate on demand password with vault
3. Add tests
4. Add/update documentation

**Questions**
1. How should this be handled in the minio-client(mc)?
-- Set output variable in minio and use that in mc

**Discussion**
1. Code in vagrant-hashistack to enable KV secret engine

### PR DESCRIPTION
**Pull request by Neha-Sinha2305**
closes #16 
- [x] Code to generate password from vault

- [x] Fix test

- [x] Documentation : I have created the following issue to continue working on the documentation [issue#21](https://github.com/fredrikhgrelland/terraform-nomad-minio/issues/21)

### CODE REVIEW

File: dev/ansible/01_generate_secrets_vault.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+- name: Generate Secrets with Vault
+  shell: |
+    export VAULT_ADDR="http://127.0.0.1:8200"
+    vault kv put secret/minio \
+    accessKey=$(vault write -field='value' gen/password length=12 symbols=0) \
+    secretKey=$(vault write -field='value' gen/password length=12 symbols=0)
+  run_once: true
```
Comment (zhenik) [inline dev/ansible/01_generate_secrets_vault.yml]: U can use `environment` 
```suggestion
- name: Generate Secrets with Vault
  shell: |
    vault kv put secret/minio \
    accessKey=$(vault write -field='value' gen/password length=12 symbols=0) \
    secretKey=$(vault write -field='value' gen/password length=12 symbols=0)
  environment:
    VAULT_ADDR: "http://127.0.0.1:8200"
```

File: example/main.tf [hunk 2]
```diff
@@ -10,12 +10,29 @@ module "minio" {
   host                            = "127.0.0.1"
   port                            = 9000
   container_image                 = "minio/minio:latest"
-  access_key                      = "minio"
-  secret_key                      = "minio123"
+  access_key                      = var.access_key
+  secret_key                      = var.secret_key
```
Comment (zhenik) [inline example/main.tf:14]: I initially thought that if variables are not set (might be bad trigger), then they are 
1. generated automatically
2. stored in the vault
3. injected in nomad job using [consul-template](https://github.com/hashicorp/consul-template#secret)

I guess I had to say that in acceptance criteria under user-story

**Review by pdmthorsrud — COMMENTED** [review]
The whole of [00-poststart-example.yml](dev/vagrant(conf/post_bootstrap/00-poststart-example.yml) should be done as part of the box (in vagrant-hashistack) IMO, because generating secrets is something a lot of modules might find handy

**Comment by zhenik** [comment]
> The whole of [00-poststart-example.yml](dev/vagrant(conf/post_bootstrap/00-poststart-example.yml) should be done as part of the box (in vagrant-hashistack) IMO, because generating secrets is something a lot of modules might find handy

**+/-** 
agree: yes, several modules will need random strings
disagree: because of coupling with specific plugin (current plugin only generates string)

Find a better way.

**Review by zhenik — APPROVED** [review]

File: example/main.tf (hunk 2 above)
Comment (claesgill) [reply example/main.tf:14]: So we can remove `var.acces_key` and `var.secret_key` or de we need them to generate the outputs?

**Comment by Neha-Sinha2305** [comment]
@pdmthorsrud : Do you think it's ok to merge these changes?

**Comment by pdmthorsrud** [comment]
> > The whole of [00-poststart-example.yml](dev/vagrant(conf/post_bootstrap/00-poststart-example.yml) should be done as part of the box (in vagrant-hashistack) IMO, because generating secrets is something a lot of modules might find handy
> 
> **+/-**
> agree: yes, several modules will need random strings
> disagree: because of coupling with specific plugin (current plugin only generates string)
> 
> Find a better way.

What do you mean because of coupling with specific plugin? Specifically that generating random strings need a plugin? If so, I still think it should be put into the box itself, not each and every module

**Comment by pdmthorsrud** [comment]
> @pdmthorsrud : Do you think it's ok to merge these changes?

Sorry late answer, it was dumb of me to request changes and then go out! I would like to resolve what's discussed above. What is your take on it?

**Comment by zhenik** [comment]
@pdmthorsrud 
- If we decide to upstream this plugin to 00-poststart-example.yml of box.  Coupling with the specific plugin: https://github.com/sethvargo/vault-secrets-gen
- generating random string need a plugin (yes - this PR)
- if we decide to use this plugin, what are our requirements (security: algorithms, support of symbols like numbers-  case- special symbols, byte strength etc...) What if we decide chose another plugin for such a needs

File: example/main.tf (hunk 2 above)
Comment (zhenik) [reply example/main.tf:14]: That depends on needs. Right now I can not answer yes or no.

**Comment by pdmthorsrud** [comment]
> @pdmthorsrud
> 
> * If we decide to upstream this plugin to 00-poststart-example.yml of box.  Coupling with the specific plugin: https://github.com/sethvargo/vault-secrets-gen
> * generating random string need a plugin (yes - this PR)
> * if we decide to use this plugin, what are our requirements (security: algorithms, support of symbols like numbers-  case- special symbols, byte strength etc...) What if we decide chose another plugin for such a needs

I'm thinking if we choose another plugin down the road we'd just change the plugin in the box. People could still implement their own plugins like what's done here, even though we supply a plugin with the box. 

Edit: If you guys don't feel it is necessary I am OK with keeping it as is right now, and changing it later if it proves useful, just so we can get this in, so that Daniel and Claes aren't waiting :)

File: example/main.tf (hunk 2 above)
Comment (zhenik) [reply example/main.tf:14]: But in the future it might be a solution, remove from user responsibility to provide credentials.

**Comment by Neha-Sinha2305** [comment]
> > @pdmthorsrud : Do you think it's ok to merge these changes?
> 
> Sorry late answer, it was dumb of me to request changes and then go out! I would like to resolve what's discussed above. What is your take on it?

Well, as long as we agree on using this plugin in this module so we could go ahead and merge this PR so that Claes and Daniel can carry on with their tasks.
We can create a separate issue where we could try to find out other plugin that could do a better job and then take a decision on whether we could add that to the template or not.

Note: The ansible tasks in 00-poststart-example.yml are idempotent so they would run just fine even if the code is duplicated in template.

**Review by pdmthorsrud — APPROVED** [review]

**Pull request merged**